Add fast diameter approximation algorithm for unweighted graphs

Rajendraprasad7 opened the pull request:
This PR implements an efficient algorithm for approximating the diameter of unweighted graphs using the Iterative Fringe Upper Bound (iFUB). 

**Reference**
Crescenzi, Pierluigi, Roberto Grossi, Leonardo Lanzi, and Andrea Marino. "A comparison of different bounds on the diameter of graphs." Theoretical Computer Science 426 (2012): 34-52.
[DOI: 10.1016/j.tcs.2012.09.018](https://doi.org/10.1016/j.tcs.2012.09.018)

**Implementation Details**

- _bfs_detailed(G, start): Performs BFS and collects level-wise information.
- _four_sweep_start_vertex(G, choice): Selects the starting node based on degree or randomly.
- _four_sweep_data(G, start_vertex): Executes the four-sweep BFS heuristic to obtain an initial lower bound.
- _ifub_helper(G, fringe_vertices): Computes eccentricity for a set of nodes.
- _ifub(G, root, lower_bound, error_threshold): Iteratively refines the diameter estimate.
- unweighted_diameter(G, error_threshold=0, start_vertex_type="degree"): The main function that computes an approximate diameter.

Rajendraprasad7 commented:
Hi @Schefflera-Arboricola, 
Could you help me with this PR? I plan on building on top of this by implementing the parallel version of it as suggested by [issue#82](https://github.com/networkx/nx-parallel/issues/82)

rossbar commented:
A couple general comments/questions:
 - Is this an approximation algorithm? If so, it should live in the `approximation` sub package!
 - How does this relate to the `extrema_bounding` approach built in to the existing distance measures?
 - Is it possible to replace the custom implementations of functions like `bfs` with the built-in networkx traversal functions? Doing so makes this a much stronger proposal as it would greatly increase the re-use of maintained & (arguably) well-tested code!

Rajendraprasad7 commented:
Thanks for replying!

**Is this an approximation algorithm?**
It could be used for approximation. The reason I did not add it to approximation was that the current approximation algorithm using 2-sweep gives a good estimate for the diameter in practice but lacks guarantees on the approximation quality. On the other hand, `iFUB` explicitly takes an `error_threshold` parameter, ensuring the additive error stays below the threshold. Setting `error_threshold = 0` gives the exact diameter.

That being said, if it makes more sense to put it under approximation, let me know, and I’ll update accordingly.

**How does this relate to the `extrema_bounding` approach built in to the existing distance measures?**

The `extrema_bounding` approach works for weighted and directed graphs, whereas `iFUB` is designed specifically for undirected and unweighted graphs. The key advantage of `iFUB` is that it is significantly faster. Here’s the performance comparison on 10 real world graphs (after extracting largest connected component):

| Graph | Nodes | Edges | Diameter | iFUB Time | extrema_bounding Time |
| ------------- | ------------- |------------- | ------------- | ------------- | ------------- |
| ca-HepPh.txt    | 11204  | 117649  | 13       | 1.6306s    | 142.9613s  |
| jazz.mtx        | 198    | 2742    | 6        | 0.0022s    | 0.0242s    |
| ca-GrQc.txt     | 4158   | 13428   | 17       | 0.1050s    | 8.3365s    |
| ca-AstroPh.txt  | 17903  | 197031  | 14       | 1.8766s    | 459.3678s  |
| rajat06.mtx     | 10922  | 28922   | 93       | 10.2895s   | 37.2680s   |
| email-Enron.txt | 33696  | 180811  | 13       | 0.5362s    | 860.9924s  |
| ca-HepTh.txt    | 8638   | 24827   | 18       | 0.3407s    | 39.8776s   |
| pkustk02.mtx    | 10800  | 410400  | 33       | 42.4671s   | 242.2084s  |
| msc10848.mtx    | 10848  | 620313  | 18       | 5.4660s    | 344.1464s  |
| Oregon-1.mtx    | 11174  | 23409   | 10       | 0.2746s    | 36.7416s   |

Further, I do not see much scope for parallelization in `extrema_bounding` whereas for `iFUB`, computing the eccentricities of fringe vertices can all be done in parallel. This fits nicely in the `embarrasingly parallel` paradigm as suggested in [issue#82](https://github.com/networkx/nx-parallel/issues/82). Contributing this algorithm to nx-parallel was my original intention.

**Is it possible to replace the custom implementations of functions like bfs with the built-in networkx traversal functions?**

- You're right, I'll try to integrate as much as possible.

Rajendraprasad7 commented:
@rossbar, @dschult , @Schefflera-Arboricola could anyone review the PR please?

dschult commented on networkx/algorithms/distance_measures.py:
```diff
@@ -1029,3 +1032,175 @@ def kemeny_constant(G, *, weight=None):
+    node
+        The selected start vertex.
+    """
+    if choice == "random":
+        start_vertex = random.choice(list(G.nodes()))
+    elif choice == "degree":
+        start_vertex = max(G.nodes(), key=lambda x: G.degree(x))
+    else:
+        raise ValueError("Invalid choice of start vertex")
```
This could be extended to allow the user to provide a function with input `G` that returns the starting node.
But that also can be done after this PR is merged. Also the syntax `key=G.degree` might be better than the `lambda`. Again, not a blocker.

dschult commented on networkx/algorithms/distance_measures.py:
```diff
@@ -1029,3 +1032,175 @@ def kemeny_constant(G, *, weight=None):
+
+@nx.utils.not_implemented_for("directed")
+def unweighted_diameter(G, error_threshold=0, start_vertex_type="degree"):
+    """Computes an approximation of the graph diameter using the IFUB algorithm.
```
The abbreviation IFUB needs to be introduced. That should probably be done in a paragraph after the first one-line description and a blank line.  Maybe something like:
```
    """Returns the diameter for an unweighted graph with optional approximation.

    This alternative to the more general `diameter` function is faster and allows
    for an approximation tolerance, though the default is to find the exact zero
    tolerance result. The function using the Iterative Fringe Upper Bound (IFUB)
    algorithm [1]_. <then describe the algorithm briefly -- with a Notes section if
    you prefer to shield the user from too many details up front.>
```

Rajendraprasad7 replied:
Fixed in latest commit.

dschult reviewed (commented):
Hello and thank you for this contribution. 

I agree with @rossbar's comments and thank you for your responses.
If the intent is to usually leave the error tolerance at 0 then this seems
good to stay in `distance_measures.py` rather than `approximations`.

We need to make it more clear why this function exists and how the user
should determine whether to use it or not. See the comment about the 
docstring's "paragraph" that describes the algorithm.

After big picture question:  why should this function exist? Is it really separate from `diameter`? Or is it a special case of diameter? Is there any advantage to having two ways to compute the diameter for unweghted diameters?

Could we just put this into the nx_parallel library as a replacement for `diameter`? That function could check for unweighted and undirected and report "can_run=False" when inputs are not right. But in the right case it could run the IFUB algorithm in parallel.  You could add the `error_threshold` keyword arg as a backend-only keyword arg too if that makes the most sense. Again, the approach of going straight into nx-parallel only works if this algorithm is a speed related replacement of the current algorithm.

Are there cases where the return value is not the same as the current NetworkX diameter? 

Minor things to take care of after the bigger points above:
- Tests: We will eventually need tests. 
- starting_node input could be generalized to be perhaps the starting node? That moves the choice of starting node outside this function -- thus providing more flexibility for the user.
- The term "vertex" is usually spelled "node" in NetworkX for uniformity.

Rajendraprasad7 commented:
Thanks for reviewing @dschult! 

**If the intent is to usually leave the error tolerance at 0 then this seems good to stay in distance_measures.py rather than approximations.**
Yes, I have added it as a backend-only keyword arg with default at 0 tolerance.

**We need to make it more clear why this function exists and how the user should determine whether to use it or not. See the comment about the docstring's "paragraph" that describes the algorithm.**
After considering your comment, I have made the function private so it is not upto the user to determine when to use it anymore(reason explained later). Fixed the docstring as suggested.

**After big picture question: why should this function exist? Is it really separate from diameter? Or is it a special case of diameter? Is there any advantage to having two ways to compute the diameter for unweghted diameters?**
It is indeed a special case of diameter i.e when the graph is undirected and unweighted. So, now the function has been made private and is invoked within the `diameter` function when a suitable graph is passed. There is no need to have two ways to compute diameter for unweighted graphs when one method is significantly faster.

**Could we just put this into the nx_parallel library as a replacement for diameter?That function could check for unweighted and undirected and report "can_run=False" when inputs are not right. But in the right case it could run the IFUB algorithm in parallel.**
Yes, thats a good idea, I'll submit a PR for nx-parallel soon. I think it is still beneficial to put this here as well, since even in its sequential form it is faster for unweighted, undirected graphs as shown in previous comment.

**Are there cases where the return value is not the same as the current NetworkX diameter?**
No, they both return the same value in all cases.

**Tests: We will eventually need tests.**
Since `iFUB` is being invoked within `diameter` function I think the current tests hold. Let me know if they have to be improved.

**The term "vertex" is usually spelled "node" in NetworkX for uniformity.**
Updated everywhere.

Rajendraprasad7 commented:
@dschult is the PR ready to be merged?

dschult commented on networkx/algorithms/distance_measures.py:
```diff
@@ -1029,3 +1033,188 @@ def kemeny_constant(G, *, weight=None):
+
+    Notes:
+    The Iterative Fringe Upper Bound (IFUB) algorithm first selects an approximate "central"
+    node through 4-sweep. A BFS tree is rooted at this node, subsequently the eccentricities
```
> first selects an approximate "central" node through 4-sweep.

What does `through 4-sweep` refer to? Is there a brief way to describe it? or a name that doesn't require a description?

Also -- can we wrap the docstring to keep it under ~80 chars? The code is enforced at <88 chars.
But docstrings linting tools aren't there yet.  Still, it'd be nice to keep it same width as the code.

Rajendraprasad7 replied:
Updated in the latest commit.

Rajendraprasad7 commented:
Function has been renamed to `_unweighted_diameter_ifub`. Fixed the ifub function call location. All helper functions have been merged. There isn't much use to them being modularised. However, when it would be implemented in nx-parallel, the only part for parallelization is 
```
cur_max_ecc = 0
    for node in fringe_vertices:
        layers = list(nx.bfs_layers(G, node))
        cur_max_ecc = max(cur_max_ecc, len(layers) - 1)
```
So it would mostly be copy pasting the entire function from here except for the fringe_vertices loop. In that case, having separate functions for each section might be useful to avoid redundant code. We could probably do that later depending on the exact need at that time.
Also, what is the purpose of `usebounds` flag?

Rajendraprasad7 commented:
@dschult If there are no other major changes to this, I'll submit a PR for the nx-parallel version of this.

Schefflera-Arboricola reviewed (commented):
I took a look at the papers referenced in the `_extrema_bounding` and `_unweighted_diameter_ifub` :

- Acc. to docstring(`Notes` section) of the `_extrema_bounding`, the `_extrema_bounding` algorithm was proposed in "[Determining the diameter of small world networks.](https://dl.acm.org/doi/abs/10.1145/2063576.2063748)". (2011)
- the algorithm proposed in this PR was proposed in "[On computing the diameter of real-world undirected graphs](https://doi.org/10.1016/j.tcs.2012.09.018)" (2013)
    - refers the `_extrema_bounding` algorithm as "TK" (Takes–Kosters algorithm)
    - `Table 5` contains comparisons of `ifub` with other algorithms (including `TK`). 
- And another paper(mentioned the docstring of `_extrema_bounding`) that proposed a similar algorithm(`SumSweep`). Paper: "[Fast diameter and radius BFS-based computation in (weakly connected) real-world graphs: With an application to the six degrees of separation games](https://doi.org/10.1016/j.tcs.2015.02.033)" (2015)
    - refers the `_extrema_bounding` algorithm as "BD" (BoundingDiameters)
    - In `Appendix C (Table C.11)` this paper compares both the `_extrema_bounding`(BD) and two variations of `iFub`(i.e. `iFub4S`(this is the one implemented in this PR, I think?) and `iFubHd`)
    - Paper's `Conclusion` states: "_We have proved that this algorithm is more general than all previous ones, and it is the first algorithm able to compute the radius and diameter of directed, not strongly connected graphs_."

Please note that the comparison in these papers are done based on the number of BFS iterations and not time. 

My conclusion after going through all of the above mentioned papers is that all of these heuristic algorithms perform better for only a specific kind(or kinds) of graphs and it doesn't make sense(to me) to have all(or multiple) of them in networkx, unless there is a strong user demand for them. But, please free to question this conclusion of mine 🙃 if you think there is a strong reason for `iFub4S`, or for that matter `SumSweep`, to be included in networkx or if `_extrema_bounding` should be replaced by it. We can also keep the `_extrema_bounding` approach in networkx and have a parallel `iFub4S` approach in nx-parallel.

Also, @Rajendraprasad7 it would be helpful if you could mention the source of the benchmarks you have shared in this comment (https://github.com/networkx/networkx/pull/7899#issuecomment-2708097647).

Thanks @Rajendraprasad7 , @dschult and @rossbar  for the discussions and work so far :) I've left a few more comments below!

Rajendraprasad7 commented:
Thanks @Schefflera-Arboricola for the detailed review! You're right, the algo this PR implements is the `iFub4S`. It only works well for special kinds of graphs which have:
- Low treewidth
- Small grids
- Diameter(G) is close to 2*Radius(G).

I wrote a script to time `iFub4S`(same code as this PR) and `TK`(imported from networkx) on 10 real world graphs. Clearly, this does not align with the results summarised in Appendix C (Table C.11) of [Fast diameter and radius BFS-based computation in (weakly connected) real-world graphs: With an application to the six degrees of separation games](https://doi.org/10.1016/j.tcs.2015.02.033). I must have made some mistake in the benchmark script, BFS count should have good correlation with run time.

I agree that having multiple heuristic algorithms for the same task is not very useful. It's better to just have the parallel `iFub4S` in nx-parallel. I'll make the corrections for the comments you've left on the code in the nx-parallel version.

dcoudert commented on networkx/algorithms/distance_measures.py:
```diff
@@ -1029,3 +1031,111 @@ def kemeny_constant(G, *, weight=None):
+    observed to compute diameters efficiently for real-world graphs [1]_.
+
+    References:
+    .. [1] Crescenzi, P., Grossi, R., Lanzi, L., & Marino, A.
+           "A comparison of different bounds on the diameter of graphs."
+           Theoretical Computer Science 426 (2012): 34-52.
```
iFUB has been first proposed in 
- P. Crescenzi, R. Grossi, C. Imbrenda, L. Lanzi, and A. Marino. Finding the Diameter in Real-World Graphs: Experimentally Turning a Lower Bound into an Upper Bound. Proceedings of 18th Annual European Symposium on Algorithms. Lecture Notes in Computer Science, vol. 6346, 302-313. Springer (2010). https://doi.org/10.1007/978-3-642-15775-2_26

The idea of using 4-sweep to find the root vertex, as done here, has been proposed in
- P. Crescenzi, R. Grossi, M. Habib, L. Lanzi, A. Marino. On computing the diameter of real-world undirected graphs. Theor. Comput. Sci. 514: 84-95 (2013).  https://doi.org/10.1016/j.tcs.2012.09.018

dcoudert reviewed (commented):
I'm happy to see an implementation of iFUB in networkx. It is such a simple algorithm (except for the 4-sweep initialization) that it should be in all graph libraries.

rossbar reviewed (commented):
This has been stale for a while, are you planning on continuing work here @Rajendraprasad7 ?
